modules/apps/layout/layout-reports-web/src/main/resources/META-INF/resources/js/LayoutReportsApp.js

```diff
@@ -12,33 +12,166 @@
  * details.
  */
 
-import React from 'react';
+import ClayAlert from '@clayui/alert';
+import ClayLoadingIndicator from '@clayui/loading-indicator';
+import {useIsMounted} from '@liferay/frontend-js-react-web';
+import {fetch} from 'frontend-js-web';
+import React, {
+	useCallback,
+	useEffect,
+	useLayoutEffect,
+	useReducer,
+} from 'react';
 
 import BasicInformation from './components/BasicInformation';
 import EmptyLayoutReports from './components/EmptyLayoutReports';
 
-export default function ({context}) {
-	const {
-		assetsPath,
-		canonicalURLs,
-		configureGooglePageSpeedURL,
-		defaultLanguageId,
-		validConnection,
-	} = context;
-
-	return (
-		<>
-			<BasicInformation
-				canonicalURLs={canonicalURLs}
-				defaultLanguageId={defaultLanguageId}
-			/>
-
-			{validConnection || (
-				<EmptyLayoutReports
-					assetsPath={assetsPath}
-					configureGooglePageSpeedURL={configureGooglePageSpeedURL}
+const initialState = {
+	data: null,
+	error: null,
+	loading: false,
+};
+
+const dataReducer = (state, action) => {
+	switch (action.type) {
+		case 'LOAD_DATA':
+			return {
+				...state,
+				loading: true,
+			};
+
+		case 'SET_ERROR':
+			return {
+				...state,
+				error: action.error,
+				loading: false,
+			};
+
+		case 'SET_DATA':
+			return {
+				data: {
+					...action.data,
+				},
+				error: action.data?.error,
+				loading: false,
+			};
+
+		default:
+			return initialState;
+	}
+};
+
+export default function ({
+	isPanelStateOpen,
+	layoutReportsDataURL,
+	portletNamespace,
+}) {
+	const isMounted = useIsMounted();
+
+	const [state, dispatch] = useReducer(dataReducer, initialState);
+
+	const safeDispatch = (action) => {
+		if (isMounted()) {
+			dispatch(action);
+		}
+	};
+
+	const getData = (fetchURL) => {
+		safeDispatch({type: 'LOAD_DATA'});
+
+		fetch(fetchURL, {
+			method: 'GET',
+		})
+			.then((response) =>
+				response.json().then((data) =>
+					safeDispatch({
+						data,
+						type: 'SET_DATA',
+					})
+				)
+			)
+			.catch(() => {
+				safeDispatch({
+					error: Liferay.Language.get('an-unexpected-error-occurred'),
+					type: 'SET_ERROR',
+				});
+			});
+	};
+
+	const handlePageAuditData = useCallback(() => {
+		getData(layoutReportsDataURL);
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [layoutReportsDataURL]);
+
+	useEffect(() => {
+		var layoutReportsPanelToggle = document.getElementById(
+			`${portletNamespace}layoutReportsPanelToggleId`
+		);
+
+		layoutReportsPanelToggle.addEventListener(
+			'mouseenter',
+			handlePageAuditData,
+			{once: true}
+		);
+		layoutReportsPanelToggle.addEventListener(
+			'focus',
+			handlePageAuditData,
+			{once: true}
+		);
```

Comment:
We shouldn't leak these event listeners. (Even though you have `once`, if there is no `mouseenter` or `focus` event we'll leak them.) The element will of course eventually be removed from the DOM on navigation, but we should still always balance `addEventListener` with `removeEventListener`, for hygiene.

You can probably do this with [`useEventListener`](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/hooks/useEventListener.ts) (it says it takes a boolean `phase` parameter, but it's not going to complain if you pass a `{once: true}` object).

Reply:
That's what I was using before need to have `{once: true}`. Thanks! I'll change it again 😄

Reply:
I change it to:
```
useEventListener(
		'mouseenter',
		handlePageAuditData,
		{once: true},
		layoutReportsPanelToggle
	);
```

And `handlePageAuditData` is calling more than once. Any hint?

Reply:
Sounds like the listener is getting added more than once (you should be able to verify that with a `console.log`). Maybe your `useEffect` is running more often than you would expect? Because if you look at the implementation of `useEventListener` it is literally [just passing those arguments straight through](https://github.com/liferay/liferay-portal/blob/70ccc851edc9ae181a6c686326553ec26b85c4cf/modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/hooks/useEventListener.ts#L35), including the `{once: true}`...